File: frontend/src/types/nodes.ts
// Node categories
export type NodeCategory = "source" | "cep-pattern" | "transform" | "sink";

// Code generation target
export type CodeTarget = "sdp" | "sss" | "sdp-or-sss";

// Every node type as a union
export type NodeType =
  // Sources (9)
  | "kafka-topic"
  | "delta-table-source"
  | "auto-loader"
  | "rest-webhook-source"
  | "cdc-stream"
  | "event-hub-kinesis"
  | "mqtt"
  | "custom-python-source"
  | "stream-simulator"
  // CEP Patterns (12)
  | "sequence-detector"
  | "absence-detector"
  | "count-threshold"
  | "velocity-detector"
  | "geofence-location"
  | "temporal-correlation"
  | "trend-detector"
  | "outlier-anomaly"
  | "session-detector"
  | "deduplication"
  | "match-recognize-sql"
  | "custom-stateful-processor"
  // Transforms (11)
  | "filter"
  | "map-select"
  | "flatten-explode"
  | "lookup-enrichment"
  | "window-aggregate"
  | "stream-stream-join"
  | "stream-static-join"
  | "union-merge"
  | "rename-cast"
  | "custom-python-udf"
  | "ml-model-endpoint"
  // Sinks (9)
  | "delta-table-sink"
  | "kafka-topic-sink"
  | "rest-webhook-sink"
  | "slack-teams-pagerduty"
  | "email-sink"
  | "sql-warehouse-sink"
  | "unity-catalog-table-sink"
  | "dead-letter-queue"
  | "lakehouse-sink";

// Config field types for dynamic form rendering
export type ConfigFieldType =
  | "text"
  | "number"
  | "select"
  | "multiselect"
  | "toggle"
  | "code"
  | "schema-picker"
  | "duration"
  | "key-value"
  | "column-picker"
  | "expression";

export interface ConfigField {
  key: string;
  label: string;
  type: ConfigFieldType;
  required: boolean;
  defaultValue?: unknown;
  placeholder?: string;
  helpText?: string;
  options?: { value: string; label: string }[];
  validation?: { min?: number; max?: number; pattern?: string };
  /** Show only when another field has a specific value (or any of values) */
  dependsOn?: { field: string; value?: unknown; values?: unknown[] };
  group?: string;
  /** For code/schema-picker: editor language (python, sql, json) */
  codeLanguage?: "python" | "sql" | "json";
}

export interface NodeDefinition {
  type: NodeType;
  label: string;
  description: string;
  category: NodeCategory;
  codeTarget: CodeTarget;
  icon: string;
  color: string;
  inputs: number;
  outputs: number;
  configFields: ConfigField[];
  advancedFields?: ConfigField[];
}

File: backend/app/models/nodes.py
"""
Node type enums and metadata for the CEP pipeline builder.

Defines all 38 supported node types with their categories and code targets.
"""

from enum import Enum
from typing import TypedDict


class NodeCategory(str, Enum):
    """Category of a pipeline node."""

    SOURCE = "source"
    CEP_PATTERN = "cep_pattern"
    TRANSFORM = "transform"
    SINK = "sink"


class CodeTarget(str, Enum):
    """Code generation target for a node type."""

    SDP = "sdp"  # Lakeflow Declarative Pipelines only
    SSS = "sss"  # Spark Structured Streaming only (CEP patterns)
    SDP_OR_SSS = "sdp_or_sss"  # Can be either


class NodeMetadata(TypedDict):
    """Metadata for a node type in the registry."""

    category: NodeCategory
    code_target: CodeTarget


# All 38 node types with their category and code_target
NODE_REGISTRY: dict[str, NodeMetadata] = {
    # Sources (8)
    "kafka-topic": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "delta-table-source": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "auto-loader": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "rest-webhook-source": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "cdc-stream": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "event-hub-kinesis": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "mqtt": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SDP_OR_SSS},
    "custom-python-source": {"category": NodeCategory.SOURCE, "code_target": CodeTarget.SSS},
    # CEP Patterns (12)
    "sequence-detector": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "absence-detector": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "count-threshold": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "velocity-detector": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "geofence-location": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "temporal-correlation": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "trend-detector": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "outlier-anomaly": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "session-detector": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "deduplication": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "match-recognize-sql": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    "custom-stateful-processor": {"category": NodeCategory.CEP_PATTERN, "code_target": CodeTarget.SSS},
    # Transforms (10)
    "filter": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "map-select": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "flatten-explode": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "lookup-enrichment": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "window-aggregate": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "stream-stream-join": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "stream-static-join": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "union-merge": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "rename-cast": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SDP_OR_SSS},
    "custom-python-udf": {"category": NodeCategory.TRANSFORM, "code_target": CodeTarget.SSS},
    # Sinks (8)
    "delta-table-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "kafka-topic-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "rest-webhook-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "slack-teams-pagerduty": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "email-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "sql-warehouse-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "unity-catalog-table-sink": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
    "dead-letter-queue": {"category": NodeCategory.SINK, "code_target": CodeTarget.SDP_OR_SSS},
}
